kornia/color/zca.py

```diff
@@ -0,0 +1,208 @@
+        r"""
+
+        Applies the whitening transform to the data
+
+        args:
+
+            x (torch.Tensor): Input data
+            include_fit (bool): Indicates whether to fit the data as part of the forward pass
+
+        returns:
+
+            torch.Tensor : The transformed data
+
+        """
+
+        if include_fit:
+            self.fit(x)
+
+        if not self.fitted:
+            raise RuntimeError("Needs to be fitted first before running. Please call fit or set include_fit to True.")
+
+        num_features: int = reduce(lambda a, b: a * b, x.size()[1::])
+
+        x_flat: torch.Tensor = torch.reshape(x, (-1, num_features))
+
+        y = (x_flat - self.mean).mm(self.T)
+
+        y = y.reshape(x.size())
+
+        return y
+
+    def inverse_transform(self, x: torch.Tensor) -> torch.Tensor:
+        r"""
+
+        Applies the inverse transform to the whitened data.
+
+        args:
+            x (torch.Tensor): Whitened data
+            include_fit (bool): Indicates whether to fit the data as part of the forward pass
+
+        returns:
+            torch.Tensor: original data
+
+
+
+        """
+
+        if not self.fitted:
+            raise RuntimeError("Needs to be fitted first before running. Please call fit or set include_fit to True.")
+
+        num_features: int = reduce(lambda a, b: a * b, x.size()[1::])
+
+        x_flat: torch.Tensor = torch.reshape(x, (-1, num_features))
+
+        y: torch.Tensor = (x_flat).mm(self.T_inv) + self.mean
+
+        y = y.reshape(x.size())
+
+        return y
+
+
+def zca_whitening_transforms(inp: torch.Tensor, eps: float = 1e-7,
+                             biased: bool = False, compute_inv: bool = False) -> Tuple[torch.Tensor, ...]:
+    r"""
+
+    Computes ZCA whitening matrix and mean vector. The output could be used in
+    :class:`~torchvision.transforms.LinearTransformation`.
+
+    See :class:`~kornia.color.ZCAWhitening` for details.
+
+
+    args:
+        inp (torch.Tensor) : input data tensor
+        eps (float) : a small number used for numerial stablility.
+        biased (bool): Whether to use the biased estimate of the covariance matrix
+        compute_inv (bool): Whether to return the inverse transform matrix
+
+
+    returns:
+        Tuple[torch.Tensor, ...]:
+        A tuple containing the ZCA matrix and the mean vector, and if compute_inv = True, the
+        inverse whitening matrix is retured as well as the final return value.
+
+
+
+    """
+
+    if not torch.is_tensor(inp):
+        raise TypeError("Input type is not a torch.Tensor. Got {}".format(
+            type(inp)))
+
+    N: int = inp.size()[0]
+    num_features: int = reduce(lambda a, b: a * b, inp.size()[1::])
+
+    mean: torch.Tensor = torch.mean(inp, dim=0, keepdim=True)
+    mean = mean.view((num_features,))
+
+    inp_center_flat: torch.Tensor = inp.view((N, num_features)) - mean
+
+    cov = inp_center_flat.t().mm(inp_center_flat)
```

Comment:
.t() won't worked in batched, same for .mm, use .transpose(-2, -1) and torch.matmul or .bmm to allow potenial broadcasting

Reply:
Maybe because I didn't mention it but the expected shape (N,D_1,D_2,...,D_n) (batch of N-D tensors). The tensors are flatten to a shape (N, D_1*D_2*...*D_n) (matrix) that's why I thought .t() or .bmm weren't needed. Hopefully this answers the next comment.